File: leetcode-codes/0410-split-array-largest-sum/solution.py
class Solution(object):
    def splitArray(self, nums, k):
        n = len(nums)
        dp=[[-1 for j in range(k+1)]for i in range(n+1)]
        prefix = [0]*(n+1)
        for i in range(1,n+1):prefix[i]=prefix[i-1]+nums[i-1]
        def dfs(i,j):
            if j==1:
                return prefix[i]
            if dp[i][j]!=-1:
                return dp[i][j]
            inf = int(1e10)
            for l in range(j-1,i):
                beforesum = prefix[i]-prefix[l]
                large = max(beforesum,dfs(l,j-1))
                inf = min(inf,large)
            dp[i][j]=inf
            return inf
        return dfs(n,k)

File: leetcode-codes/0410-split-array-largest-sum/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int splitArray(vector<int>& nums, int k) {
        int l=0,r=accumulate(nums.begin(),nums.end(),0);
        int res = 2e9;
        while(l<=r){
            int mid = (l+r)>>1;
            int sum=0;
            int cnt =1;
            int _max =0;
            for(int i =0;i<nums.size();i++){
                if(sum+nums[i]>mid){
                    _max = max(_max,sum);
                    sum=0;
                    cnt++;
                }
                sum+=nums[i];
            }
            _max = max(sum,_max);
            if(cnt<=k){
                r=mid-1;
                res = min(res,_max);
            }
            else{
                l= mid+1;
            }
        }
        return res;
    }
};
